packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -0,0 +1,209 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'package:flutter/foundation.dart';
+
+import 'basic.dart';
+import 'editable_text.dart';
+import 'framework.dart';
+
+// TODO(justinmc): Autocomplete should make it easy to do asynchronous searches.
+/// A type for autocomplete search functions.
+///
+/// [AutocompleteController] uses a search function to search through
+/// [AutocompleteController.options] and return a subset as results.
+typedef AutocompleteSearchFunction<T> = List<T> Function(String query);
+
+/// A type for indicating the selection of an autocomplete result.
+typedef OnSelectedAutocomplete<T> = void Function(T result);
+
+/// A builder for the selectable results given the current autocomplete query.
+typedef AutocompleteResultsBuilder<T> = Widget Function(
+  BuildContext context,
+  List<T> results,
+  OnSelectedAutocomplete<T> onSelected,
+);
+
+/// A builder for the query field in autocomplete.
+typedef AutocompleteFieldBuilder = Widget Function(
+  BuildContext context,
+  TextEditingController textEditingController,
+);
+
+/// A controller for the [AutocompleteCore] widget.
+///
+/// Can also be used in a stand-alone manner to implement autocomplete behavior
+/// in a fully custom UI.
+@immutable
+class AutocompleteController<T> {
+  /// Create an instance of AutocompleteController.
+  AutocompleteController({
+    this.options,
+    this.search,
+    TextEditingController textEditingController,
+  }) : assert(search != null || options != null, "If a search function isn't specified, Autocomplete will search by string on the given options."),
+       textEditingController = textEditingController ?? TextEditingController() {
+    this.textEditingController.addListener(_onQueryChanged);
+  }
+
+  /// All possible options that can be searched.
+  ///
+  /// If left null, a custom [search] method must be provided that handles the
+  /// options to be searched on its own.
+  final List<T> options;
+
+  /// The [TextEditingController] that represents the query.
+  final TextEditingController textEditingController;
+
+  /// A search function that takes some [options] and returns a subset as
+  /// results.
+  ///
+  /// If [options] is null, then this field must not be null. This may be the
+  /// case when querying an external service for search results, for example.
+  ///
+  /// Defaults to a simple string-matching search of [options].
+  final AutocompleteSearchFunction<T> search;
```

Comment:
Yeah I think you're right.  This function may sometimes get more complicated than a simple filter, such as when calling a backend endpoint, but I think that it will always have the ability to return multiple results, which is more like a filter than a search.  I'll change the naming.